File: C/tcp-chat/tcp_chat_cli.c
#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <strings.h>
#include <string.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <sys/ioctl.h>
#include <netdb.h>

#define BUF_SIZE 300
#define SERVER_PORT "9999"

// read a string from stdin protecting buffer overflow
#define GETS(B,S) {fgets(B,S-2,stdin);B[strlen(B)-1]=0;}

int main(int argc, char **argv) {
	int err, sock;
	unsigned char lsize;
	char nick[BUF_SIZE], linha[BUF_SIZE], buff[BUF_SIZE];
	struct addrinfo  req, *list;
	fd_set rfds;


	if(argc!=2) {
        	puts("Server IPv4/IPv6 address or DNS name is required as argument");
        	exit(1); }

	bzero((char *)&req,sizeof(req));
	// let getaddrinfo set the family depending on the supplied server address
	req.ai_family = AF_UNSPEC;
	req.ai_socktype = SOCK_STREAM;
	err=getaddrinfo(argv[1], SERVER_PORT , &req, &list);
	if(err) {
        	printf("Failed to get server address, error: %s\n",gai_strerror(err)); exit(1); }

	sock=socket(list->ai_family,list->ai_socktype,list->ai_protocol);
	if(sock==-1) {
        	perror("Failed to open socket"); freeaddrinfo(list); exit(1);}

	if(connect(sock,(struct sockaddr *)list->ai_addr, list->ai_addrlen)==-1) {
        	perror("Failed connect"); freeaddrinfo(list); close(sock); exit(1);}

	freeaddrinfo(list);

	printf("Connected, enter nickname: ");GETS(nick,BUF_SIZE);

	for(;;) {
        	FD_ZERO(&rfds);
        	FD_SET(0,&rfds); FD_SET(sock,&rfds);
        	select(sock+1,&rfds,NULL,NULL,NULL);
        	if(FD_ISSET(0,&rfds)) {
                	GETS(linha,BUF_SIZE);
                	if(!strcmp(linha,"exit")) {
                        	lsize=0;
                        	write(sock,&lsize,1);
                        	read(sock,&lsize,1);
                        	break;
                        	}
                	sprintf(buff,"(%s) %s",nick,linha);
                	lsize=strlen(buff);
                	write(sock,&lsize,1);
               		write(sock,buff,lsize);
                	}
        	if(FD_ISSET(sock,&rfds)) {
                	read(sock,&lsize,sizeof(lsize));
                	read(sock,buff,lsize);
			buff[lsize]=0;
                	puts(buff);
                	}
      		}
	close(sock);
	exit(0);
	}

File: JAVA/tcp-chat/TcpChatCli.java
import java.io.*; 
import java.net.*; 

class TcpChatCli {    
	static InetAddress serverIP;
	static Socket sock;


	public static void main(String args[]) throws Exception {
		String nick, frase;
		byte[] data = new byte[300];

		if(args.length!=1) {
			System.out.println(
				"Server IPv4/IPv6 address or DNS name is required as argument");
			System.exit(1); }

		try { serverIP = InetAddress.getByName(args[0]); }
		catch(UnknownHostException ex) {
   			System.out.println("Invalid server: " + args[0]);
    			System.exit(1); }

		try { sock = new Socket(serverIP, 9999); }
		catch(IOException ex) {
   			System.out.println("Failed to connect.");
   			System.exit(1); }

		BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
		DataOutputStream sOut = new DataOutputStream(sock.getOutputStream());

   		System.out.println("Connected to server");
		System.out.print("Enter your nickname: "); nick = in.readLine();

		// start a thread to read incoming messages from the server
		Thread serverConn = new Thread(new TcpChatCliConn(sock));
		serverConn.start();


		while(true) { // read messages from the console and send them to the server
			frase=in.readLine();
			if(frase.compareTo("exit")==0) 
				{ sOut.write(0); break;}
			frase="(" + nick + ") " + frase;
			data = frase.getBytes();
			sOut.write((byte)frase.length());
			sOut.write(data,0,(byte)frase.length());
			}

		serverConn.join();
		sock.close();
		} 
	}



class TcpChatCliConn implements Runnable {    
	private Socket s;
	private DataInputStream sIn;

	public TcpChatCliConn(Socket tcp_s) { s=tcp_s;}

	public void run() {
		int nChars;
		byte[] data = new byte[300];
		String frase;

		try {
			sIn = new DataInputStream(s.getInputStream());
			while(true) {
				nChars=sIn.read();
				if(nChars==0) break;
				sIn.read(data,0,nChars);
				frase = new String(data, 0, nChars);
				System.out.println(frase);
				}
			}
		catch(IOException ex) { System.out.println("Client disconnected."); }
		}
	}




